include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,283 @@
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+#define GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+
+#include "glow/Base/Type.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Node.h"
+#include "glow/Support/Error.h"
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include <functional>
+
+namespace glow {
+
+///===---------------------------------------------------------------------===//
+///                               SplitNodeOption
+///===---------------------------------------------------------------------===//
+/// Generic split node option as base type. All the split options refer to the
+/// split procedure performed with respect to the tensor associated with the
+/// output operand of a node. If a node has multiple output operands then the
+/// option commonly refers to the first output operand.
+class SplitNodeOption {
+
+  /// All the dimensions of a tensor used for splitting in increasing order.
+  std::vector<size_t> splitDims_;
+
+public:
+  /// Ctor.
+  SplitNodeOption(llvm::ArrayRef<size_t> splitDims) : splitDims_(splitDims) {}
+
+  /// \returns the split dims.
+  llvm::ArrayRef<size_t> getSplitDims() const { return splitDims_; }
+
+  /// \returns the index of \p splitDim in the \ref splitDims_ array.
+  size_t getSplitDimIdx(dim_t splitDim) const;
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  virtual std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const = 0;
+
+  /// Dtor.
+  virtual ~SplitNodeOption() = default;
+};
+
+/// Split option used for splitting a node along the dimensions \ref splitDims
+/// with the given number of chunks \ref numChunks for each dimension. Since the
+/// split does not always result in equal chunks, you can choose to start the
+/// splitting with the bigger chunks first (one unit bigger than the others) by
+/// using the flag \ref bigChunksFirst. The number of chunks for each dimension
+/// must be lower than the respective dimension size, otherwise error is thrown.
+class SplitNodeByNumChunks : public SplitNodeOption {
+
+  /// Number of chunks for each split dimension.
+  std::vector<dim_t> numChunks_;
+
+  /// Whether to start splitting with bigger chunks first.
+  bool bigChunksFirst_{true};
+
+public:
+  /// Ctor.
+  SplitNodeByNumChunks(llvm::ArrayRef<size_t> splitDims,
+                       llvm::ArrayRef<dim_t> numChunks,
+                       bool bigChunksFirst = true)
+      : SplitNodeOption(splitDims), numChunks_(numChunks),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK_EQ(splitDims.size(), numChunks.size())
+        << "Mismatch between 'splitDims' and 'numChunks' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
```

Comment:
Should the return type be a mutable vector? If not, return an ArrayRef.

Reply:
Mutable vector so I left it `std::vector`.